src/gcp/cloudscheduler.ts

```diff
@@ -12,38 +14,117 @@ export interface Schedule {
+    maxBackoffDuration?: string;
+    maxDoublings?: number;
+  };
 }
 
-export function createJob(schedule: Schedule): Promise<void> {
+/**
+ * Creates a cloudScheduler job.
+ * If another job with that name already exists, this will return a 409.
+ * @param job The job to create.
+ */
+export function createJob(job: Job): Promise<any> {
   // the replace below removes the portion of the schedule name after the last /
   // ie: projects/my-proj/locations/us-central1/jobs/firebase-schedule-func-us-east1 would become
   // projects/my-proj/locations/us-central1/jobs
-  const strippedName = schedule.name.replace(/\/[^\/]+$/, "");
+  const strippedName = job.name.substring(0, job.name.lastIndexOf("/"));
   return api.request("POST", `/${VERSION}/${strippedName}`, {
     auth: true,
     origin: api.cloudschedulerOrigin,
-    data: Object.assign({ timeZone: "America/Los_Angeles" }, schedule),
+    data: Object.assign({ timeZone: DEFAULT_TIME_ZONE }, job),
   });
 }
 
-export function deleteJob(name: string): Promise<void> {
+/**
+ * Deletes a cloudScheduler job with the given name.
+ * Returns a 404 if no job with that name exists.
+ * @param name The name of the job to delete.
+ */
+export function deleteJob(name: string): Promise<any> {
   return api.request("DELETE", `/${VERSION}/${name}`, {
     auth: true,
     origin: api.cloudschedulerOrigin,
   });
 }
 
-export async function createOrReplaceJob(schedule: Schedule): Promise<void> {
-  const jobName = `${schedule.name.split("/")[5]}`;
+/**
+ * Gets a cloudScheduler job with the given name.
+ * If no job with that name exists, this will return a 404.
+ * @param name The name of the job to get.
+ */
+export function getJob(name: string): Promise<any> {
+  return api.request("GET", `/${VERSION}/${name}`, {
+    auth: true,
+    origin: api.cloudschedulerOrigin,
+  });
+}
+
+/**
+ * Updates a cloudScheduler job.
+ * Returns a 404 if no job with that name exists.
+ * @param job A job to update.
+ */
+export function updateJob(job: Job): Promise<any> {
+  // Note that name cannot be updated.
+  return api.request("PATCH", `/${VERSION}/${job.name}`, {
+    auth: true,
+    origin: api.cloudschedulerOrigin,
+    data: Object.assign({ timeZone: DEFAULT_TIME_ZONE }, job),
+  });
+}
+
+/**
+ * Checks for a existing job with the given name.
+ * If none is found, it creates a new job.
+ * If one is found, and it is identical to the job parameter, it does nothing.
+ * Otherwise, if one is found and it is different from the job param, it updates the job.
+ * @param job A job to check for and create, replace, or leave as appropriate.
+ * @throws { FirebaseError } if an error response other than 404 is received on the GET call.
+ */
+export async function createOrReplaceJob(job: Job): Promise<any> {
+  const jobName = `${job.name.split("/").pop()}`;
   try {
-    await createJob(schedule);
-    logLabeledSuccess("functions", `created scheduler job ${jobName}`);
+    const existingJob = await getJob(job.name);
```

Comment:
Ah! It took me a while to figure out why this `try` was here.

In the [api](https://github.com/firebase/firebase-tools/blob/master/src/api.js#L180) source, there's an option called `resolveOnHTTPError`. That can be set to `true` and the API will return the [normal response stuff](https://github.com/firebase/firebase-tools/blob/master/src/api.js#L57) which might be a little easier to quickly check the status code and keep moving.

The `try/catch` block here is a little tricky to follow. AND, actually, if `updateJob` below causes an error, it will be caught and (in the horrible case it's a `404` error) cause `createJob` to be called as well! I think I'd definitely recommend using `resolveOnHTTPError` to help with this.